auth.go

```
@@ -0,0 +1,218 @@
+	authd.Pubkey = m.keystore.PublicPEM()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain device public key")
+	}
+
+	tentok, err := m.store.ReadAll(authTenantTokenName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read tenant token")
+	}
+	log.Debugf("tenant token: %s", tentok)
+
+	// fill tenant token
+	authd.TenantToken = string(tentok)
+
+	// fetch sequence number
+	num, err := m.seqnum.Get()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain sequence number")
+	}
+	authd.SeqNumber = num
+
+	log.Debugf("authorization data: %v", authd)
+
+	databuf := &bytes.Buffer{}
+	enc := json.NewEncoder(databuf)
+
+	err = enc.Encode(&authd)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to encode auth request")
+	}
+
+	reqdata := databuf.Bytes()
+
+	// generate signature
+	sig, err := m.keystore.Sign(reqdata)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to sign auth request")
+	}
+
+	return &AuthRequest{
+		Data:      reqdata,
+		Code:      AuthCode(tentok),
+		Signature: sig,
+	}, nil
+}
+
+func (m *MenderAuthManager) RecvAuthResponse(data []byte) error {
+	if len(data) == 0 {
+		return errors.New("empty auth response data")
+	}
+
+	if err := m.store.WriteAll(authTokenName, data); err != nil {
+		return errors.Wrapf(err, "failed to save auth token")
+	}
+	return nil
+}
+
+func (m *MenderAuthManager) AuthCode() (AuthCode, error) {
+	data, err := m.store.ReadAll(authTokenName)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return noAuthCode, nil
+		}
+		return noAuthCode, errors.Wrapf(err, "failed to read auth token data")
+	}
+
+	return AuthCode(data), nil
+}
+
+func (m *MenderAuthManager) HasKey() bool {
+	if m.keystore.Private() != nil {
+		return true
+	}
+	return false
+}
+
+func (m *MenderAuthManager) GenerateKey() error {
+	if err := m.keystore.Generate(); err != nil {
+		log.Errorf("failed to generate device key: %v", err)
+		return errors.Wrapf(err, "failed to generate device key")
```

Ideally we should propagate the error up to a place that it has enough context to be meaningful. In practice, people either log everything or virtually nothing. Perhaps feedback from testing could be useful here.